syz-manager/rpc.go

```diff
@@ -262,161 +173,156 @@ func (serv *RPCServer) Check(a *rpctype.CheckArgs, r *int) error {
 		return nil
 	}
-	newMaxSignal := serv.maxSignal.Diff(a.MaxSignal.Deserialize())
-	if !newMaxSignal.Empty() {
-		serv.maxSignal.Merge(newMaxSignal)
-		serv.stats.maxSignal.set(len(serv.maxSignal))
-		for _, f1 := range serv.fuzzers {
-			if f1 == f || f1.rotated {
-				continue
-			}
-			f1.newMaxSignal.Merge(newMaxSignal)
+	// Note: ReportGenerator is already initialized if coverFilter is enabled.
+	rg, err := getReportGenerator(serv.cfg, serv.modules)
+	if err != nil {
+		return err
+	}
+	filtered := 0
+	for _, pc := range pcs {
+		if serv.coverFilter[uint32(rg.RestorePC(pc))] != 0 {
+			filtered++
 		}
 	}
-	if f.rotated {
-		// Let rotated VMs run in isolation, don't send them anything.
+	serv.stats.corpusCoverFiltered.add(filtered)
+	return nil
+}
+
+func (serv *RPCServer) shutdownInstance(name string) []byte {
+	var runner *Runner
+	if val, _ := serv.runners.LoadAndDelete(name); val != nil {
+		runner = val.(*Runner)
+	} else {
 		return nil
 	}
-	r.MaxSignal = f.newMaxSignal.Split(2000).Serialize()
-	if a.NeedCandidates {
-		r.Candidates = serv.mgr.candidateBatch(serv.batchSize)
-	}
-	if len(r.Candidates) == 0 {
-		batchSize := serv.batchSize
-		// When the fuzzer starts, it pumps the whole corpus.
-		// If we do it using the final batchSize, it can be very slow
-		// (batch of size 6 can take more than 10 mins for 50K corpus and slow kernel).
-		// So use a larger batch initially (we use no stats as approximation of initial pump).
-		const initialBatch = 50
-		if len(a.Stats) == 0 && batchSize < initialBatch {
-			batchSize = initialBatch
-		}
-		for i := 0; i < batchSize && len(f.inputs) > 0; i++ {
-			last := len(f.inputs) - 1
-			r.NewInputs = append(r.NewInputs, f.inputs[last])
-			f.inputs[last] = rpctype.Input{}
-			f.inputs = f.inputs[:last]
-		}
-		if len(f.inputs) == 0 {
-			f.inputs = nil
-		}
+
+	runner.mu.Lock()
+	if runner.requests == nil {
+		// We are supposed to invoke this code only once.
+		panic("Runner.requests is already nil")
 	}
-	for _, inp := range r.NewInputs {
-		inp.Cover, inp.Signal = f.instModules.Decanonicalize(inp.Cover, inp.Signal)
+	oldRequests := runner.requests
+	runner.requests = nil
+	runner.mu.Unlock()
+
+	// If the object does not exist, there would be no oldRequests either.
+	fuzzerObj := serv.mgr.getFuzzer()
+	for _, req := range oldRequests {
+		// The VM likely crashed, so let's tell pkg/fuzzer to abort the affected jobs.
+		// TODO: distinguish between real VM crashes and regular VM restarts?
+		fuzzerObj.Done(req, &fuzzer.Result{Stop: true})
 	}
-	log.Logf(4, "poll from %v: candidates=%v inputs=%v maxsignal=%v",
-		a.Name, len(r.Candidates), len(r.NewInputs), len(r.MaxSignal.Elems))
-	return nil
+	return runner.machineInfo
 }
 
-func (serv *RPCServer) shutdownInstance(name string) []byte {
-	serv.mu.Lock()
-	defer serv.mu.Unlock()
+func (serv *RPCServer) distributeSignalDelta(plus, minus signal.Signal) {
+	serv.runners.Range(func(key, value any) bool {
+		if value == nil {
```

Comment:
Do we store nil's there? Or it's some artifact of sync.Map?

Reply:
It was an artifact of code rewriting :) It's not needed, I've dropped it.